### PR TITLE
feat(review-workflows): add stage attribute in entities

### PR DESCRIPTION
**Pull request by nathan-pichon**
### What does it do?

In this PR, there are multiple additions:
- Possibility to include an `reviewWorkflows` options in any content-type
- A new stage attribute in the entities where reviewWorkflows is activated
- A migration of entities when reviewWorkflows is being activated

### How to test it?

You can start your Strapi project with the admin panel. Go into the content type builder, activate review workflow on any content-type.
Then it should migrate all entities inside this content-type and add a relation to the first stage of the default workflow.
You can retrieve this stage on any entity in the content-type by populating `strapi_reviewWorkflows_stage`.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

File: packages/core/admin/ee/server/services/review-workflows/review-workflows.js [hunk 1]
```diff
@@ -24,31 +37,111 @@ function buildStagesConnectArray(stages) {
...
+      // Execute a raw SQL query to insert records into the join table mapping the specified content type with the first stage of the default workflow.
+      // Only entities that do not have a record in the join table yet are selected.
+      await strapi.db.connection
+        .raw(`INSERT INTO ${joinTable.name} (${idColumn.name}, field, "order", ${joinTable.joinColumn.name}, ${typeColumn.name})
+                SELECT
+                    entity.id as ${idColumn.name},
+                    '${ENTITY_STAGE_ATTRIBUTE}' as field,
+                    1 as "order",
+                    ${firstStage.id} as ${joinTable.joinColumn.name},
+                    '${contentTypeUID}' as ${typeColumn.name}
```
Comment (jhoward1994) [inline packages/core/admin/ee/server/services/review-workflows/review-workflows.js:118]: I don't think we need to specify aliases for these selections. The query should work like this as we specify the order in the insert into

```
entity.id,
'${ENTITY_STAGE_ATTRIBUTE}',
1,
${firstStage.id},
'${contentTypeUID}'
```

**Review by jhoward1994 — APPROVED** [review]

**Comment by jhoward1994** [comment]
@Marc-Roig I'm wondering what you think of these changes? everything is working well for me other than one minor comment. If you agree I'll commit that and merge this PR

File: packages/core/admin/ee/server/services/review-workflows/review-workflows.js [hunk 2]
```diff
@@ -24,31 +37,111 @@ function buildStagesConnectArray(stages) {
...
+                FROM ${contentTypeMetadata.tableName} entity
+                LEFT JOIN ${joinTable.name} jointable
+                ON  entity.id = jointable.${idColumn.name}
+                AND jointable.${typeColumn.name} = '${contentTypeUID}'
+                WHERE jointable.${idColumn.name} IS NULL`);
+    };
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/review-workflows.js:124]: For future PRs :) I would do this in a knex query, also, adding API tests  would ensure this works for all dbs.

⚠️  I have not tried this query, but it should be something like this.
```js
const con = strapi.db.connection;
const selectStatement = con
  .select(
    `entity.id as ${idColumn.name}`,
    `${ENTITY_STAGE_ATTRIBUTE} as field`,
    con.raw(`?? as order`, [1]), // TODO : Add workflow id
    `${firstStage.id} as ${joinTable.joinColumn.name}`,
    `${contentTypeUID} as ${typeColumn.name}`
  )
  .leftJoin(`${joinTable.name} jointable`, function() {
      this
        .on('entity.id', '=', `jointable.${idColumn.name}`)
        .andOn(`jointable.${typeColumn.name}`, '=',  `${contentTypeUID}`)
  })
  .where(`jointable.${idColumn.name}`, null)
  .from(joinTable.name)
  .toSQL();

// Insert the clone relations
await this.createQueryBuilder(joinTable.name)
  .insert(
    con.raw(
      `(${columnsToInsert.join(',')})  ${selectStatement.sql}`,
      selectStatement.bindings
    )
  )
  .execute();
```

**Review by Marc-Roig — APPROVED** [review]
Added a couple of comments, but I think we can work on them on other PRs :) 
@jhoward1994

File: packages/core/admin/ee/server/services/review-workflows/review-workflows.js (hunk 1 above)
Comment (jhoward1994) [reply packages/core/admin/ee/server/services/review-workflows/review-workflows.js:118]: yep good point 🙂

File: packages/core/admin/ee/server/services/review-workflows/review-workflows.js [hunk 3]
```diff
@@ -24,31 +37,111 @@ function buildStagesConnectArray(stages) {
...
+    const up = async (contentTypeUID) => {
+      const contentTypeMetadata = strapi.db.metadata.get(contentTypeUID);
+      const { target, morphBy } = contentTypeMetadata.attributes[ENTITY_STAGE_ATTRIBUTE];
+      const { joinTable } = strapi.db.metadata.get(target).attributes[morphBy];
+      const { idColumn, typeColumn } = joinTable.morphColumn;
+
+      // Execute a raw SQL query to insert records into the join table mapping the specified content type with the first stage of the default workflow.
+      // Only entities that do not have a record in the join table yet are selected.
+      await strapi.db.connection
+        .raw(`INSERT INTO ${joinTable.name} (${idColumn.name}, field, "order", ${joinTable.joinColumn.name}, ${typeColumn.name})
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/review-workflows.js:112]: in mysql and mariadb, "order" should be enclosed using backticks and not double quotes :/

**Comment by jhoward1994** [comment]
See https://github.com/strapi/strapi/pull/16020#issuecomment-1461663776

File: packages/core/admin/ee/server/services/review-workflows/review-workflows.js [hunk 4]
```diff
@@ -24,31 +37,111 @@ function buildStagesConnectArray(stages) {
...
+  return async ({ contentTypes }) => {
+    // TODO To be refactored when multiple workflows are added
+    const defaultWorkflow = await strapi
+      .query(WORKFLOW_MODEL_UID)
+      .findOne({ populate: ['stages'] });
+
+    // This is possible if this is the first start of EE, there won't be any workflow in DB before bootstrap
+    if (!defaultWorkflow) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (gu-stav) [reply packages/core/admin/ee/server/services/review-workflows/review-workflows.js:101]: I think that was the idea yes: if the user doesn't have any workflows (yet) create the first one upon feature-activation (in EE).

[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/admin/ee/server/services/review-workflows/review-workflows.js:101]: then in this case, on the first startup, the entities will not be assigned any default stage.

**Pull request merged**